PDB pre-processing

joaomcteixeira opened the pull request:
* Implements a first idea for preprocessing PDB files to check and edit them to the specifications of HADDOCK3.
* The new `gear` is independent and can be used by activating the `preprocess=true` general parameter key. It is `false` by default.
* In case the PDBs are preprocessed, a copy of the original files will be kept in the `data` folder inside the run directory
* users can also use the `haddock-pp` client to process PDBs before executing haddock3.
* Importantly, this PR cleans many of the CNS-related topology files and better integrates them with the Python shell. For example, the different kinds of supported molecules can now be imported, and their nature and parameters can be used throughout the python code.

Other pull requests will likely follow this one with improvements as our experience using this feature also improves. For example, making this `gear` an actual module. For this, the [PDB preprocessing project](https://github.com/haddocking/haddock3/projects/16) exists to organise related issues and pull requests.

More information at #143

amjjbonvin commented on src/haddock/gear/preprocessing.py:
```diff
@@ -0,0 +1,153 @@
+        partial(pdb_rplresname.run, name_from='MSE', name_to='MET'),
+        partial(pdb_fixinsert.run, option_list=[]),
+        #
+        pdb_tidy.run,
```
Does pdb_tidy adds TER statements at each chain break? E.g. also when part of a loop it missing?  Or only at the end of a chain?  (the latter is the correct behaviour for haddock purposes otherwise, depending on the setting you might introduce charged termini in locations where they should not be).

joaomcteixeira replied:
`pdb_tidy` adds `TER` in chain breaks. So maybe, `tidy` is not the best here. Maybe we need to write a local "tidy".

joaomcteixeira replied:
`pdb_tidy -strict`

amjjbonvin commented:
Don’t we have a tool that only adds TER statements at the end of a chain, but not within a chain?

joaomcteixeira commented:
`pdb_tidy -strict` you are right :wink:

joaomcteixeira commented on src/haddock/gear/preprocessing.py:
```diff
@@ -0,0 +1,239 @@
+    line_by_line_processing_steps = [
+        pdb_keepcoord.run,
+        pdb_selaltloc.run,
+        partial(pdb_rplresname.run, name_from='MSE', name_to='MET'),
```
And MSE -> MET (with HETATM changed to ATOM)

joaomcteixeira commented on tests/data/good.pdb:
```diff
@@ -0,0 +1,741 @@
+ATOM      2  CA  GLY A   2      25.630   2.325  57.666  1.00 38.61      A    C  
+ATOM      3  C   GLY A   2      26.653   3.346  58.106  1.00 51.20      A    C  
+ATOM      4  O   GLY A   2      26.838   4.422  57.516  1.00 51.80      A    O  
+ATOM      6  CA  ASP A   3      28.334   3.616  59.916  1.00 45.09      A    C  
```
Yeah! just missed the `NA`. I am taking only backbone atoms for testing for simplicity. Do I need to take also hydrogens to test the preprocessing steps?

rvhonorato replied:
Don't think we need the hydrogens, but would be good to have the CB, haddock can rebuild the side-chains but only if there's a carbon-beta.

joaomcteixeira replied:
Okay, this PDB is just for testing the sanitizing, but that could indeed be a test to make sure at least CB is built. I will update that :+1:

amjjbonvin commented:
Haddock can rebuild in principle anything :-) but having the cb gives you a direction for the rebuild

rvhonorato commented:
> Haddock can rebuild in principle anything :-) but having the cb gives you a direction for the rebuild

I learn something new about haddock everyday! 🤓

rvhonorato commented on src/haddock/core/supported_molecules.py:
```diff
@@ -1,10 +1,46 @@
-https://bianca.science.uu.nl/haddock2.4/library
-
-
-Write decent documentation .-)
```
😆

joaomcteixeira replied:
Made the general `.top` file reader as you suggested :+1:. Ended up using just a small regex for the atom lines, else did with python.

rvhonorato replied:
🥲

rvhonorato commented on src/haddock/clis/cli_pp.py:
```diff
@@ -0,0 +1,53 @@
+
+ap.add_argument('pdb_files', help="Input PDB files.", nargs='+')
+ap.add_argument('-d', '--dry', help="Perform a dry run.", action="store_true")
+ap.add_argument('-t', '--topfile', help="Additional .top files.", nargs="*")
```
This is nice, but for integrating this into other things (the servers), would be nice if there was an option that was not additive; `Exclusive .top files`? 

It could be that the list of supported residues is (becomes) different between services and then this pre-processing could only be used if whatever other application using it contains also supports all residues that are supported by haddock3 - which might not be the case.

joaomcteixeira replied:
I will look carefully to make it as decoupled as possible from the allowed residues. But the implementation already is to some extent. Parameters can be given `None` to ignore them. And I can make it to pass a list of residues also, to avoid the `top` terminology.

Remember it is also possible to use the preprocessing from the library interface, which allows additional fine-tunning compared to the client.

amjjbonvin commented:
Conflicts need to be resolved, but further I am fine with merging it - we should simply start testing / using it and will see if changes are needed

joaomcteixeira commented:
Yes. I am finishing #425 and #392. After I will review this one and we merge. But, I agree we need to merge this one to start testing it.

joaomcteixeira commented:
This feature is almost ready. It required much more work than initially expected. Among other minor things, I am waiting on https://github.com/haddocking/pdb-tools/pull/136.

This `preprocessing` feature does several corrections that sometimes may be desired or not. For example,

1. `pdb_reres` may break compatibility with the `.tbl` files.
2. when doing scoring for CAPRI events, we do input PDBs where each `MODEL` may have its own labels. Currently, there's a check that avoids these cases.

A possibility is to make `preprocessing` a module with a parameter for each correcting action that is `True` by default and that could be defined as `False`. For example:

```toml
[preprocess]
reres = false

[topoaa]
# (...)
```

Is this too much for now? Should we continue with the current state that performs all actions and can be activated or deactivated with `preprocess = False/True` in the general section of config file?

amjjbonvin commented:
Rather than `pdb_reres` which looses for example gap information, I strongly recommend to use `pdb_shift` instead.

And the pre-processing module will be useful, but users should be warned / instructed to create their `tbl` files based on the numbering of the pre-processed files.

joaomcteixeira commented:
Adding `preprocessing` as a module makes sense if considering enabling and disabling certain functionalities. Not clear which could be now. Because implementing the `preprocessing` is a project on its own (https://github.com/haddocking/haddock3/projects/16), we can merge now and continue working on it as we test it.

joaomcteixeira commented:
I have finished implementing this feature (the first step of this big project). Please re-read the original message in this PR for updates. You can access the documentation at `reference -> gear -> preprocessing` and `command-line clients -> PDB processing client` and have a look to see if it is well explained.

joaomcteixeira commented:
@mgiulini a question since you are testing the client:
Currently, the processed PDBs are saved in the same folder as the input PDBs. Is it okay that way, or is it best to save the processed PDBs in the current working directory?

mgiulini commented on src/haddock/clis/cli_pp.py:
```diff
@@ -0,0 +1,106 @@
+    cli(ap, main)
+
+
+def main(*pdb_files, dry=False, topfile=None, suffix=SUFFIX_DEFAULT):
+    """Process PDB files."""
+    new_residues = read_additional_residues(topfile) if topfile else None
+
+    processed_pdbs = process_pdbs(
+        *pdb_files,
+        dry=dry,
+        user_supported_residues=new_residues,
+        )
+
+    out_files = add_suffix_to_files(pdb_files, suffix)
+    save_lines_to_files(out_files, processed_pdbs)
+
+    return
```
I would insert a couple of log calls here. Especially a small message at the beginning, so that the user knows things are getting done. given that dealing with large ensembles can take a while, it's good to know haddock3-pp is working properly.

joaomcteixeira replied:
Done, thanks.

mgiulini reviewed (commented):
good addition!
Of course I didn't have the time to check all the code, but I added a couple of comments that imo could improve the user and developer experience.

On a more high-level perspective:
- Can you maybe add a .cfg file where the preprocessing is performed?
- I tried to test the cli on some CAPRI-54 ensembles (for example Target192), and I always received the following error message: 
`haddock.gear.preprocessing.ModelsDifferError: Labels in MODEL 2 differ from MODEL 1.`
Is this the expected behavior?

joaomcteixeira commented:
Okay. Made a new option `-odir` to save the output to a specific folder. Defaults to the CWD.

joaomcteixeira commented on src/haddock/gear/preprocessing.py:
```diff
@@ -0,0 +1,1059 @@
+
+    chain_ids = []
+    new_structures = []
+    for lines in structures:
```
Why? Each item of `structures` is something that contains the `lines`.

joaomcteixeira commented:
Regarding the CAPRI-54 example, yes, that is the expected behaviour. In normal docking runs, all models should have the same labels (i.e., be the same protein). For the CAPRI case, i.e. only evaluating, the `preprocessing` in its current state should not be used. UNLESS, we define a parameter to deactivate some of the checks, which is something I would leave progressively to other pull requests. One of the proposals is to make this a module by itself. Currently, `preprocessing` does all checks or none. The `preprocess` is deactivated by default.

joaomcteixeira commented on src/haddock/gear/preprocessing.py:
```diff
@@ -0,0 +1,1059 @@
+#. replace ``HSE`` to ``HIS``
+#. replace ``HID`` to ``HIS``
+#. replace ``HIE`` to ``HIS``
+#. add charges to ions
+#. add_charges_to_ions, see :py:func:`add_charges_to_ions`
```
Thanks

mgiulini commented:
> Regarding the CAPRI-54 example, yes, that is the expected behaviour. In normal docking runs, all models should have the same labels (i.e., be the same protein). For the CAPRI case, i.e. only evaluating, the `preprocessing` in its current state should not be used. UNLESS, we define a parameter to deactivate some of the checks, which is something I would leave progressively to other pull requests. One of the proposals is to make this a module by itself. Currently, `preprocessing` does all checks or none. The `preprocess` is deactivated by default.

Sounds reasonable, thanks!

amjjbonvin commented:
In CAPRI, one target usually consists of the same protein, but not all groups might model the same sequence… 

There are still checks that need to be done, e.g. the histidine naming to correct HSD and HSE AMBER-like naming (not official PDB).  The only check that should not be done is to ensure the consistency of the models.

This is also true for scoring. And even more as in HADDOCK3 in principle we can give different proteins or ligand as one ensemble.

> Regarding the CAPRI-54 example, yes, that is the expected behaviour. In normal docking runs, all models should have the same labels (i.e., be the same protein). For the CAPRI case, i.e. only evaluating, the preprocessing in its current state should not be used. UNLESS, we define a parameter to deactivate some of the checks, which is something I would leave progressively to other pull requests. One of the proposals is to make this a module by itself. Currently, preprocessing does all checks or none. The preprocess is deactivated by default.
>

joaomcteixeira commented:
Indeed. I suggest using another PR to work on turning specific checks into options. What do you think?

joaomcteixeira commented:
> Can you maybe add a .cfg file where the preprocessing is performed?

 I could try to break one example and then use the option to correct it. But since we are at the beginning of working on this feature, could we live it also for another PR, so we think more clearly about it? Thanks!

amjjbonvin commented:
> Indeed. I suggest using another PR to work on turning specific checks into options. What do you think?
> 
Sounds reasonable.

The consistency of model is the most critical check which we should be able to turn off.
Without that we are loosing some of the new stuff we can now do with haddock3

amjjbonvin commented:
The scoring test example in haddock2.4 would be a good one to use also since it consists of very different complexes.

joaomcteixeira commented:
I think I have addressed all issues. Note this project Kanban stores the progress in the preprocessing project: https://github.com/haddocking/haddock3/projects/16

joaomcteixeira commented:
Merged. :relaxed: 
Please add all experience or feedback using this feature to new issues. Thanks!
